kratos/spaces/ublas_space.h

```diff
@@ -723,18 +729,32 @@ class UblasSpace
         IndexPartition(system_size).for_each([&](std::size_t Index){
             bool empty = true;
 
-            const std::size_t col_begin = Arow_indices[Index];
-            const std::size_t col_end = Arow_indices[Index + 1];
+            const std::size_t col_begin = r_Arow_indices[Index];
+            const std::size_t col_end = r_Arow_indices[Index + 1];
 
             for (std::size_t j = col_begin; j < col_end; ++j) {
-                if(std::abs(Avalues[j]) > zero_tolerance) {
+                if(std::abs(r_Avalues[j]) > zero_tolerance) {
                     empty = false;
                     break;
                 }
             }
 
             if(empty) {
-                rA(Index, Index) = scale_factor;
+                const auto it_Acol_indices_row_begin = it_Acol_indices_begin + col_begin;
+                const auto it_Acol_indices_row_end = it_Acol_indices_begin + col_end;
+
+                const auto lower = std::lower_bound(it_Acol_indices_row_begin, it_Acol_indices_row_end, Index);
+                const auto upper = std::upper_bound(it_Acol_indices_row_begin, it_Acol_indices_row_end, Index);
```

Comment:
```suggestion
                const auto upper = std::upper_bound(lower, it_Acol_indices_row_end, Index);
```


sry for the pickiness, but if we're doing it, we may as well do it properly.


No need to start searching from the range begin again. lower_bound <= upper_bound always

Reply:
i don't think you need the upper (in line with the suggestion of @matekelemen):

~~~cpp
if(empty) {
                const auto it_Acol_indices_row_begin = it_Acol_indices_begin + col_begin;
                const auto it_Acol_indices_row_end = it_Acol_indices_begin + col_end;

                const auto lower = std::lower_bound(it_Acol_indices_row_begin, it_Acol_indices_row_end, Index);
                if (lower != it_Acol_indices_row_end && *lower==j) { // Index was found
                    r_Avalues[std::distance(it_Acol_indices_begin, lower)] = scale_factor;
                } else {
            #ifdef KRATOS_DEBUG
                    KRATOS_ERROR << "Diagonal term (" << Index << ", " << Index << ") is not defined in the system matrix" << std::endl;
            #else
~~~

per regards the error i am not really sure. 

This s a scaling operation. If the row is not there than i would say it makes no sense to scale it ... not that this is an error ... 
What is actually wrong in my opinion is to put a nonzero diagonal where there was no diagonal (in a sense equivalent to having a zero diagonal)